### DISCUSSION TITLE
Notes Jul 13th, 2022

### DISCUSSION THREAD

**Discussion by wipfli** [maplibre/maplibre#69] [Technical Steering Committee Meetings]
## Organizational Questions
@ntadej: Can we have https://docs.maplibre.org/? This may allow us to setup nicer URLs for it as https://maplibre.org/maplibre-gl-js-docs/style-spec/ looks a bit strange. Or maybe rather https://maplibre.org/docs/maplibre-gl-js/style-spec/
* The subdomain docs.maplibre.org can only point to one GitHub repo.
* Let's keep it how it is right now because it works well with GitHub Pages.

@wipfli Who would like to be on the Governing Board?
* Work is unpaid.
* How much time does it take? Variable. At least 4 meetings per year.
* Board has the final word on money.
* How can one apply? Governing Board will draft a process.
* Can non-Voting Members be on the Governing Board? [GitHub Discussion.](https://github.com/maplibre/maplibre/discussions/32#discussioncomment-3143429)

@lseelenbinder Voting procedure updates
* https://github.com/maplibre/maplibre/discussions/32 
* Have 35 Voting Members now.
* Might create a repo for Voting Members who might become members of a GitHub team.

@maxammann publication of Digitalfabrik Support blog post: https://github.com/maplibre/maplibre.github.io/pull/55
* Promotions for sponsors are not published yet. It is important that everyone is on the same page and that we formalize what sponsors get.
* Concern that the blog post is too much promotion for Digitalfabrik.
* Meaningful content on the blog about how people use MapLibre is good.
* We could open up the blog to community members such that they can publish interesting things there. [GitHub Discussion.](https://github.com/maplibre/maplibre/discussions/67)

@maxammann Order stickers/shirts for FOSS4G? :)
* Design is on Google Drive.
* @klokan, @nyurik, @mojodna, @maxammann, @wipfli are going.
* People will pay directly Max if they want a t-shirt.

Next meeting: Aug 10th, 2022

## MapLibre GL JS Technical Questions
@wipfli Experimental features in the style spec https://github.com/maplibre/maplibre-gl-js/pull/1191
* Keep back-compatibility and do not remove features once you introduce new features in the style spec.
* Introducing a flag for experimental features is an option.

@HarelM What should we do about the shaders diverging from native, e.g.:
* https://github.com/maplibre/maplibre-gl-js/pull/1244

@HarelM Better understand what’s currently needed in the maplibre-gl-js repository for the native in order to prevent further breaking changes (Last example: serialization and test data)
* GL JS shaders, style spec, render test fixtures, query test fixtures are used in native.
* Aim for consistency between GL JS and Native even if there is no feature parity.

## MapLibre GL Native Technical Questions
@ntadej & @wipfli: Versioning? Currently core is versioned independently of individual SDK. At least Qt could stay in sync, but for Qt we plan to bump to 2.0 due to renaming.
* Plan to tag 2.0 version soon, ideally before shaders update. -> @roblabs mentioned problems with OpenGL on iOS 16 or so. Let's hold back with shader updates until we addressed those. (edit (roblabs):  There are several [blocking issues related to iOS 15.5](https://github.com/maplibre/maplibre-gl-native/issues?q=is%3Aissue+is%3Aopen+iOS+15.5) that should be resolved before we split the repo.
* Do we want to support 32-bit Windows (hint: it does not compile currently)? (not discussed)

@wipfli Changlog entries

@wipfli mapbox maptiler api keys https://github.com/maplibre/maplibre-gl-native/issues/349
* The mapbox:// protocol can be removed from native. Note that telemetry and tracking was already removed.

@ntadej: https://github.com/maplibre/qt-geoservices-maplibre-gl almost ready

@wipfli: Warm welcome to @rinigus, @nvanfleet, and @brendan-ward who have now write access to Native!

@wipfli Secrets like firebase and gradle keys are not documented and we don't know how they are managed. This blocks CI improvements and future releases. https://github.com/maplibre/maplibre-gl-native/issues/312. Introduce key manager? https://github.com/maplibre/maplibre/issues/63
* Secrets might be at MapTiler because the repo was migrated from there. @klokan will check.

At the end of the meeting, we had a nice discussion about graphics layer stuff. Here are some points:
* Keep moving MapLibre GL JS forward even if it increases technical debt on Native. In particular, accept changes to the shaders and to the style spec in GL JS.
* Use MapLibre sponsorship funds to pay people to work on Native. Highest priority seems to be parity with GL JS. A goal is to update the submodule and use the latest shaders.
* Aim for consistency between Native and GL JS. Even if Native lags in features, let's keep the two projects coupled and develop them together.
* Shaders are shared between GL JS and Native at the moment. With Metal implementation in Native, this might change. WebGPU might be an alternative shading language for the future which could compile to OpenGL and Metal if I understood correctly. Browser adoption has to be in place though first.

## Attendees (@GitHub handles)
@HarelM
@wipfli
@lseelenbinder
@mojodna
@thehoneymad
@maxammann
@Joxit
@hy9be
@raincityfun
@nyurik

Feel free to share your thoughts, ask any questions, or add things which I forgot... Thanks!

**Comment by maxammann** [comment]
> @ntadej: Can we have https://docs.maplibre.org/? This may allow us to setup nicer URLs for it as https://maplibre.org/maplibre-gl-js-docs/style-spec/ looks a bit strange. Or maybe rather https://maplibre.org/docs/maplibre-gl-js/style-spec/
> 
>     * The subdomain docs.maplibre.org can only point to one GitHub repo.
> 
>     * Let's keep it how it is right now because it works well with GitHub Pages.

I don't really understand the argument against it. We are currently publishing from several repositories to `maplibre.org`. We can do the same with `docs.maplibre.org`. I think we should absolutely do this, actually also as a security feature. I don't really like it that `maplibre-rs` is able to publish to https://maplibre.org/maplibre-rs/book/.

**Reply by wipfli** [reply]
The subdomain https://docs.maplibre.org would forward to https://maplibre.github.io? Or how would it exactly work?

**Reply by wipfli** [reply]
https://maplibre.org goes to https://maplibre.github.io

**Reply by ntadej** [reply]
Why not make then https://maplibre-docs.github.io with also nicely structured landing page that would summarise all the available docs?

**Reply by wipfli** [reply]
But that would be a different github org, right?

**Reply by ntadej** [reply]
But any repo can have its github pages published. And it does not really matter what the GitHub URL for it is, what matter is the subdomain you link to it. Or am I missing something?

**Reply by ntadej** [reply]
Ah, I get it now. We use the default GitHub subfolders and don't really push to one repo.

**Reply by wipfli** [reply]
Yeah

**Reply by wipfli** [reply]
I am happy to intoduce the docs subdomain as long as we can have the gl js docs and others in separate github repos

**Reply by Joxit** [reply]
Hi there.

Any repositories will have a default github pages named `<username or org>.github.io`. It will be the only one domain name from github. 

You can configure a default custom domain name by creating a repository named `<username or org>.github.io` and attach your domain name written in a file `CNAME` eg: https://github.com/Joxit/joxit.github.io set joxit.dev as default domain and all my pages are under Joxit.dev

Each repository can have it own custom domain name if there is a `CNAME` file eg: https://github.com/Joxit/helm-charts set helm.joxit.dev as domain name.

But I think for the last example, all the code may be under the same repository (all documentation must be written here)

**Reply by Joxit** [reply]
Ha ha I'm too slow 😂

**Reply by maxammann** [reply]
Hmhm, maybe its not as easy as I thought. Maybe its time for some non-github web hosting :D

If we stick to a single organisation and want to deploy from multiple repositories, then we'd have to use different subdomains:

- docs.gl-js.maplibre.org
- docs.gl-native.maplibre.org
- docs.rs.maplibre.org

**Reply by Joxit** [reply]
Or use github action in all projects, build documentation and push to the doc repository with something like https://github.com/marketplace/actions/push-directory-to-another-repository

**Reply by HarelM** [reply]
While I understand the beauty in the suggested URLs I wondering when was the last time you actually typed a full URL instead of either using the browser's auto complete, Google search or a link somewhere...
If someone is extremely passionate about doing this, then by all means please do, but from the looks of it we are struggling to keep the docs up-to-date, working, and with latest technologies so it feels to me like we should focus there instead of trying to find a solution to something that is OK. My two cents at least, but it's a free country :-)

**Reply by maxammann** [reply]
I agree, that there are probably more important topics. Having documentation at all is more important than where it is, but: Changing the URL schema later on is not good. That would case a lot of broken links.

So ideally the URL schema should not depend on the technical environment.

I like Github Pages for simple stuff, but outsourcing web hosting could be beneficial for maplibre.

**Reply by maxammann** [reply]
As we are using now cloudflare, I would suggest to move from Github Pages to Cloudflare Pages. That way we would have maximum flexibility an also have all our hosting "under one roof". (And have IPv6 support, Github apparently hates IPv6... 😡 )

**Reply by nyurik** [reply]
If possible, I would prefer to use github to the max, not because it is so good while cloudflare is worse, but to keep it simple. Also, github makes it much easier for anyone to try the same system in their own forks, whereas with CF contributors won't be able to experiment with the build on their own.  At some point the benefits of some external system will outweigh this concern and we will start using them, but the fewer deps like this we have, the better.

**Reply by ntadej** [reply]
I'm indifferent, but many other static site generators/hosts provide actions to have a deployment per MR which we currently do not have. That would be useful to have for example to publish test results that generate webpages (e.g. rendering tests).

**Reply by wipfli** [reply]
I think GitHub Pages solves our problem of serving static websites at the moment, or am I missing something?

**Reply by ntadej** [reply]
At the moment, certainly.

But having something more flexible would allow us to do more things.

**Reply by wipfli** [reply]
Ah OK, what could be interesting stuff to do with Cloudflare?

**Reply by wipfli** [reply]
For storing artifacts from CI runs I think we can use GitHub, or AWS S3, or Cloudflare R2 (I think that is what they call it...). But artifacts seem a different concern than hosting static sites.

**Reply by nyurik** [reply]
BTW, if we haven't done this already, we should probably [configure github static pages](https://docs.github.com/en/pages/configuring-a-custom-domain-for-your-github-pages-site/managing-a-custom-domain-for-your-github-pages-site) to use maplibre.org domain.

**Reply by maxammann** [reply]
We are already using the maplibre.org domain on Github.

So, I think Github Pages does its job right now well. But there are already a few usecases which are not possible with Github Pages which would be nice:

* Using subdomains for maplibre documentation: See the discussion above. Having stable links for at least a few years is absolutely beneficial imo.
* Setting custom HTTP headers for maplibe-rs demos.

That are two usecases arleady which are not possible through Github Pages. Github Pages is nice for simple setups, but MapLibre is above that state now. More and more projects are coming to MapLibre which all have different requirements.

At the same time, relying completely on Github, a single entity, also comes with some risk. We already use:

* Github Issues
* Github Repos (Git)
* Github Pull Requests
* Github Actions
* and Github Pages.

I think we should try to become not solely dependant on Github as an organisation.


I'm happy with using Github Pages for experiements. But anything more sophisticated than an experiment could be hosted on a proper "product". I'm not necessarily voting for Cloudflare, any would do the job. But Cloudflare is what we have now :D

**Reply by roblabs** [reply]
> For storing artifacts from CI runs I think we can use GitHub, or AWS S3, or Cloudflare R2 (I think that is what they call it...). But artifacts seem a different concern than hosting static sites.

I recommend GitHub for storing artifacts & assets. It's already in place and works like a champ.  In fact, the iOS Framework binary asset is already [hosted on GitHub under releases](https://github.com/maplibre/maplibre-gl-native/releases/tag/ios-v5.12.2).  FYI — The GitHub uses S3 for hosting the assets, so hopefully we won't "*move from S3 to S3*". 😃

**Comment by roblabs** [comment]
My understanding of how to do subdomains on GitHub with GH Pages would be as follows:

1.  If you had a site like https://docs-for-all.maplibre.org, then you should have a repo in MapLibre called `docs-for-all`. 
2.  Then configure your host records so that the host name `docs-for-all` has the address `maplibre.github.io` (see 2nd image for example.  While configuring you can use `dig` to verify if it should resolve.

**Reply by maxammann** [reply]
Exactly, but we can not deploy from any repository in the Github organisation to docs-for-all.maplibre.org.

That is the misunderstanding I had in the beginning, and we concluded that this usecase it not easily posisble with Github Pages.

**Comment by klokan** [comment]
What pain is there with current domain setup or documentation?

What is your suggested improvement of that pain?

Please be precise on WHAT you want to reach, the HOW (GH pages, CF pages,
... is just implementation).

- www.maplibre.org points to maplibre.github.io already
- any repo @maplibre has it own www.maplibre.org/repo/ website if you
create gh-pages in that repo

If you want maplibre.org/docs/ or docs.maplibre.org for shared
documentation for all projects it is immediately doable with current setup.

CloudFlare already accelerates our GH-Pages via their CDN. I believe IP6 is
already enabled.

**Reply by maxammann** [reply]
> What pain is there with current domain setup or documentation?

Like written above, the following pain points exist:


- Using subdomains for maplibre documentation. Having stable links for at least a few years is absolutely beneficial imo. Examples: docs.maptiler.com, docs.python.org, docs.mapbox.org
- Setting custom HTTP headers

> What is your suggested improvement of that pain?

Using a more sophisticated web hosting would allow us to deploy all documentation to a single subdomain (e.g. docs.maplibre.org). **If we continue to use Github Pages, then this is not possible. Github Pages is incompatible with multiple repositories pushing to a separate subdomain like docs.maplibre.org**.

> Please be precise on WHAT you want to reach, the HOW (GH pages, CF pages,... is just implementation).

I actually did that. I just repeated what I already wrote in this comment.

> If you want maplibre.org/docs/ or docs.maplibre.org for shared documentation for all projects it is immediately doable with current setup.

I think this is the root of our misunderstanding here. The docs.maplibre.org usecase is NOT possible. If it is possible then I maybe misunderstood something and we are totally fine with Github Pages for now :)

> CloudFlare already accelerates our GH-Pages via their CDN. I believe IP6 is already enabled.

👍🏼

**Comment by roblabs** [comment]
We are in agreement on this.  At this point, I recommend we improve our documentation before considering moving to a subdomain or another provider.  I do not recommend moving anything to S3 at this time (there was discussion above), as GitHub assets are already S3.  Thanks.

> *If you want maplibre.org/docs/ or docs.maplibre.org for shared
documentation for all projects it is immediately doable with current setup.*

**Reply by maxammann** [reply]
> If you want maplibre.org/docs/ or docs.maplibre.org for shared
documentation for all projects it is immediately doable with current setup.

I'm fairly certain that having docs.maplibre.org for shared documentation is NOT possible. If it is then let's keep everything like it is and switch to that solution I would say (I can also do it 😃)

**Reply by wipfli** [reply]
We could create a repository at https://github.com/maplibre/docs, serve GitHub Pages from its main branch, and configure DNS to point https://docs.maplibre.org to https://maplibre.github.io/docs. Then we could push the static page output of for example the MapLibre GL JS documentation to this new the maplibre/docs repo maybe into a `/maplibre-gl-js` folder. This is possible with Personal Access Tokens (PAT) which are tied to one person.

**Reply by wipfli** [reply]
You need to use the PAT approach because GitHub inter-repo automation does not really exist...

**Reply by wipfli** [reply]
I guess for security reasons. Like workflows cannot trigger workflows etc....

**Reply by maxammann** [reply]
Yes that should work! But at that point it gets fairly complicated already. 

Some user will have to geberate the PAT though afaik. So it will be bound to a user. I would not want to CI to use my PAT for deploymet.
PAT also should have an expiration date. Therefore, the  secret must be rotated.

Definitely possible though, but in my opinion already more complicated than using a dedicated web hoster.

**Reply by wipfli** [reply]
I agree...

**Reply by maxammann** [reply]
The best workaround to get a PAT is to create a new Github user and give it rights to write to the docs repository. Then generate a PAT from it and use it. I did this in other projects, but that way we have to maintain additional secrets and the whole deployment gets quite confusing, because it is a defacto workaround.

**Reply by HarelM** [reply]
From my point of view the problem is not technical but more of a process problem:
I believe the best method (which is not for all repos but worth looking into) is to have the docs as part of the repo, so if you change the code you change the docs and they are coupled (in a good way).
If you split the docs from their repos you create maintenance havoc, IMHO.

**Reply by maxammann** [reply]
True, thats maybe another issue to look into (moving gl-js docs to gl-js repo). The issue discussed above is only about deployment. Not about where the docs live.

**Reply by Joxit** [reply]
If you don't want to use a PAT, you can 
1. create the project  https://github.com/maplibre/docs and do the DNS configuration
2. create a [schedule workflow](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#schedule) that will, every x days:
    1. clone every projects (PAT not needed since all projects are public)
    2. for each projects in defined folders build the documentation
    3. use the action [`actions/deploy-pages@main`](https://github.com/actions/deploy-pages) and deploy the website (FYI I'm using this action [here](https://github.com/jawg/booking-map-example/blob/main/.github/workflows/pages.yml)

=> No need PAT, Cloudflare, S3....

In any case, for SEO it will be very complicated to re-index everything if we use a subdomain :)

**Reply by maxammann** [reply]
That's quite a nice solution, except that it might mix up concerns slightly. That way the ci in docs can get quite complicated. The ci scripts will depend on tools required by gl-js, gl-native, MapLibre-rs etc.

Also then the CI scripts will no longer be next to the documentation sources if I understood correctly. They would be in the docs repo.

**Reply by Joxit** [reply]
Or... We can use the example ["Using an action inside a different private repository than the workflow"](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-using-an-action-inside-a-different-private-repository-than-the-workflow) except this time it will be a public repository (no need of PAT) :smile: 

With this, every workflow will be in their own repositories and we will call them in docs :wink: 

Is that good for you ?

**Reply by maxammann** [reply]
Good idea! Unfortunately, I think its only possible to share actions, and not jobs which include for example the definiton of the operating system in the CI.

A usecase for private repos could also arise in the future.

I feel like we are trying to work around features of a proprietary platform here. If everyone really wants to stick to Github Pages then I would say lets keep it like that. And go with the solution @Joxit proposed, which is the best workaround I've seen :)

**Reply by HarelM** [reply]
Can someone please define a proper problem statement?? I honestly don't understand what this discussion is telling to solve...
I'm not talking about futuristic "maybes"...
"The docs should be in a single address" is a solution, not a problem statement...

**Reply by maxammann** [reply]
There are a few low-prio problems with the current setup.

* We already mentioned that docs.maplibre.org may look more "pleasant" to some eyes (whatever that means :D)
* Security: any repository seems to be able to publish directly to maplibre.org. That means its possible to change links to downloads/text on the main website.
* Separation of concerns: maplibre.org and the docs website have different goals. Also if for some reason we habe a page: https://maplibre.org/maplibre page in our website at some point, we are no longer able to publish docs from the maplibre repo. There would be a conflict between the deployment of the website and docs. So we have to make manually sure that that there wont be issues.

There are probably more. From a sysadmin perspective it makes defintely sense to deploy different services do different domains.

**Reply by HarelM** [reply]
1. Arguable...
2. If everyone that needs to publish the documentation related to their package need access, the security problem exists anyway, or there's a bottleneck in terms of approval of the documentation updates.
3. If we don't use docs subdomain this problem doesn't exist as every repo has its own route, assuming we don't do anything fancy in maplibre.io repo.

In any case, this thread seems too long and it might be better to talk about it in our next steering committee meeting.

**Reply by maxammann** [reply]
Yeah lets establish maybe a working group for IT instrastructure topics and avoid discussing such topics in the large group. I think we can not reach any conclusions that way. @HarelM can you take care of this topic during the next meeting?

**Reply by HarelM** [reply]
Sure :-)
